File: api/src/test/java/org/pmiops/workbench/api/DataDictionaryTest.java
package org.pmiops.workbench.api;

import static com.google.common.truth.Truth.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.pmiops.workbench.SpringTest;
import org.pmiops.workbench.access.AccessTierService;
import org.pmiops.workbench.cdr.CdrVersionMapper;
import org.pmiops.workbench.cdr.CdrVersionService;
import org.pmiops.workbench.cdr.ConceptBigQueryService;
import org.pmiops.workbench.cdr.dao.DSDataDictionaryDao;
import org.pmiops.workbench.cdr.model.DbDSDataDictionary;
import org.pmiops.workbench.cohortbuilder.CohortQueryBuilder;
import org.pmiops.workbench.cohorts.CohortService;
import org.pmiops.workbench.conceptset.ConceptSetService;
import org.pmiops.workbench.conceptset.mapper.ConceptSetMapper;
import org.pmiops.workbench.config.CdrBigQuerySchemaConfigService;
import org.pmiops.workbench.dataset.BigQueryTableInfo;
import org.pmiops.workbench.dataset.DataSetServiceImpl;
import org.pmiops.workbench.dataset.mapper.DataSetMapperImpl;
import org.pmiops.workbench.db.dao.AccessTierDao;
import org.pmiops.workbench.db.dao.CdrVersionDao;
import org.pmiops.workbench.db.model.DbCdrVersion;
import org.pmiops.workbench.exceptions.BadRequestException;
import org.pmiops.workbench.exceptions.NotFoundException;
import org.pmiops.workbench.firecloud.FireCloudService;
import org.pmiops.workbench.genomics.GenomicExtractionService;
import org.pmiops.workbench.model.DataDictionaryEntry;
import org.pmiops.workbench.model.Domain;
import org.pmiops.workbench.notebooks.NotebooksService;
import org.pmiops.workbench.utils.TestMockFactory;
import org.pmiops.workbench.utils.mappers.CommonMappers;
import org.pmiops.workbench.workspaces.WorkspaceAuthService;
import org.pmiops.workbench.workspaces.WorkspaceService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.boot.test.context.TestConfiguration;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.context.annotation.Import;
import org.springframework.test.annotation.DirtiesContext;

@DataJpaTest
@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
public class DataDictionaryTest extends SpringTest {

  @Autowired private AccessTierDao accessTierDao;
  @Autowired private CdrVersionDao cdrVersionDao;
  @Autowired private DSDataDictionaryDao dsDataDictionaryDao;
  @Autowired private DataSetController dataSetController;

  private static DbCdrVersion cdrVersion;

  @TestConfiguration
  @Import({
    CommonMappers.class,
    DataSetController.class,
    DataSetServiceImpl.class,
    DataSetMapperImpl.class,
    CdrVersionService.class,
  })
  @MockBean({
    BigQueryService.class,
    CdrBigQuerySchemaConfigService.class,
    CohortService.class,
    CohortQueryBuilder.class,
    ConceptBigQueryService.class,
    ConceptSetService.class,
    ConceptSetMapper.class,
    FireCloudService.class,
    NotebooksService.class,
    WorkspaceService.class,
    WorkspaceAuthService.class,
    AccessTierService.class,
    CdrVersionMapper.class,
    GenomicExtractionService.class
  })
  static class Configuration {}

  @BeforeEach
  public void setUp() {
    cdrVersion = TestMockFactory.createDefaultCdrVersion(cdrVersionDao, accessTierDao);

    DbDSDataDictionary dbDSDataDictionary = new DbDSDataDictionary();
    dbDSDataDictionary.setRelevantOmopTable(BigQueryTableInfo.getTableName(Domain.DRUG));
    dbDSDataDictionary.setFieldName("TEST FIELD");
    dbDSDataDictionary.setOmopCdmStandardOrCustomField("A");
    dbDSDataDictionary.setDescription("B");
    dbDSDataDictionary.setFieldType("C");
    dbDSDataDictionary.setDataProvenance("D");
    dbDSDataDictionary.setSourcePpiModule("E");
    dbDSDataDictionary.setDomain("Drug");

    dsDataDictionaryDao.save(dbDSDataDictionary);
  }

  @Test
  public void testGetDataDictionaryEntry() {
    final Domain domain = Domain.DRUG;
    final String domainValue = "FIELD NAME / DOMAIN VALUE";

    DbDSDataDictionary dbDSDataDictionary = new DbDSDataDictionary();
    dbDSDataDictionary.setRelevantOmopTable(BigQueryTableInfo.getTableName(domain));
    dbDSDataDictionary.setFieldName(domainValue);
    dbDSDataDictionary.setOmopCdmStandardOrCustomField("A");
    dbDSDataDictionary.setDescription("B");
    dbDSDataDictionary.setFieldType("C");
    dbDSDataDictionary.setDataProvenance("D");
    dbDSDataDictionary.setSourcePpiModule("E");
    dbDSDataDictionary.setDomain("DRUG");
    dsDataDictionaryDao.save(dbDSDataDictionary);

    DataDictionaryEntry response =
        dataSetController
            .getDataDictionaryEntry(cdrVersion.getCdrVersionId(), domain.toString(), domainValue)
            .getBody();

    assertThat(response.getRelevantOmopTable())
        .isEqualTo(dbDSDataDictionary.getRelevantOmopTable());
    assertThat(response.getFieldName()).isEqualTo(dbDSDataDictionary.getFieldName());
    assertThat(response.getOmopCdmStandardOrCustomField())
        .isEqualTo(dbDSDataDictionary.getOmopCdmStandardOrCustomField());
    assertThat(response.getDescription()).isEqualTo(dbDSDataDictionary.getDescription());
    assertThat(response.getFieldType()).isEqualTo(dbDSDataDictionary.getFieldType());
    assertThat(response.getDataProvenance()).isEqualTo(dbDSDataDictionary.getDataProvenance());
    assertThat(response.getSourcePpiModule()).isEqualTo(dbDSDataDictionary.getSourcePpiModule());
  }

  @Test
  public void testGetDataDictionaryEntry_invalidCdr() {
    assertThrows(
        BadRequestException.class,
        () -> dataSetController.getDataDictionaryEntry(-1L, Domain.DRUG.toString(), "TEST FIELD"),
        "Invalid CDR Version");
  }

  @Test
  public void testGetDataDictionaryEntry_invalidDomain() {
    assertThrows(
        BadRequestException.class,
        () ->
            dataSetController.getDataDictionaryEntry(
                cdrVersionDao.findAll().iterator().next().getCdrVersionId(),
                "random",
                "TEST FIELD"));
  }

  @Test
  public void testGetDataDictionaryEntry_notFound() {
    assertThrows(
        NotFoundException.class,
        () -> dataSetController.getDataDictionaryEntry(1L, Domain.DRUG.toString(), "random"));
  }
}

File: api/src/test/java/org/pmiops/workbench/api/DataDictionaryTest.kt
package org.pmiops.workbench.api

import org.assertj.core.api.Assertions.assertThat

import java.sql.Timestamp
import java.time.Clock
import java.time.Instant
import java.time.ZoneId
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.rules.ExpectedException
import org.junit.runner.RunWith
import org.pmiops.workbench.cdr.dao.ConceptDao
import org.pmiops.workbench.dataset.DataSetMapper
import org.pmiops.workbench.dataset.DataSetMapperImpl
import org.pmiops.workbench.db.dao.CdrVersionDao
import org.pmiops.workbench.db.dao.CohortDao
import org.pmiops.workbench.db.dao.ConceptSetDao
import org.pmiops.workbench.db.dao.DataDictionaryEntryDao
import org.pmiops.workbench.db.dao.DataSetDao
import org.pmiops.workbench.db.dao.DataSetService
import org.pmiops.workbench.db.model.CdrVersion
import org.pmiops.workbench.db.model.DataDictionaryEntry
import org.pmiops.workbench.exceptions.BadRequestException
import org.pmiops.workbench.exceptions.NotFoundException
import org.pmiops.workbench.firecloud.FireCloudService
import org.pmiops.workbench.model.Domain
import org.pmiops.workbench.notebooks.NotebooksService
import org.pmiops.workbench.test.FakeClock
import org.pmiops.workbench.workspaces.WorkspaceService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.liquibase.LiquibaseAutoConfiguration
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.boot.test.context.TestConfiguration
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Import
import org.springframework.test.context.junit4.SpringRunner

@RunWith(SpringRunner::class)
@DataJpaTest
@Import(LiquibaseAutoConfiguration::class)
class DataDictionaryTest {

    @Autowired
    internal var bigQueryService: BigQueryService? = null
    @Autowired
    internal var cdrVersionDao: CdrVersionDao? = null
    @Autowired
    internal var cohortDao: CohortDao? = null
    @Autowired
    internal var conceptDao: ConceptDao? = null
    @Autowired
    internal var conceptSetDao: ConceptSetDao? = null
    @Autowired
    internal var dataDictionaryEntryDao: DataDictionaryEntryDao? = null
    @Autowired
    internal var dataSetDao: DataSetDao? = null
    @Autowired
    internal var dataSetMapper: DataSetMapper? = null
    @Autowired
    internal var dataSetService: DataSetService? = null
    @Autowired
    internal var fireCloudService: FireCloudService? = null
    @Autowired
    internal var notebooksService: NotebooksService? = null
    @Autowired
    internal var workspaceService: WorkspaceService? = null

    @Autowired
    internal var dataSetController: DataSetController? = null

    @Rule
    var expectedEx = ExpectedException.none()

    @TestConfiguration
    @Import(DataSetController::class, DataSetMapperImpl::class)
    @MockBean(BigQueryService::class, CohortDao::class, ConceptDao::class, ConceptSetDao::class, DataSetDao::class, DataSetService::class, FireCloudService::class, NotebooksService::class, WorkspaceService::class)
    internal class Configuration {
        @Bean
        fun clock(): Clock {
            return CLOCK
        }
    }

    @Before
    fun setUp() {
        val cdrVersion = CdrVersion()
        cdrVersionDao!!.save(cdrVersion)

        val dataDictionaryEntry = DataDictionaryEntry()
        dataDictionaryEntry.cdrVersion = cdrVersion
        dataDictionaryEntry.definedTime = Timestamp(CLOCK.millis())
        dataDictionaryEntry.relevantOmopTable = ConceptSetDao.DOMAIN_TO_TABLE_NAME[Domain.DRUG]
        dataDictionaryEntry.fieldName = "TEST FIELD"
        dataDictionaryEntry.omopCdmStandardOrCustomField = "A"
        dataDictionaryEntry.description = "B"
        dataDictionaryEntry.fieldType = "C"
        dataDictionaryEntry.dataProvenance = "D"
        dataDictionaryEntry.sourcePpiModule = "E"
        dataDictionaryEntry.transformedByRegisteredTierPrivacyMethods = true

        dataDictionaryEntryDao!!.save(dataDictionaryEntry)
    }

    @Test
    fun testGetDataDictionaryEntry() {
        val domain = Domain.DRUG
        val domainValue = "FIELD NAME / DOMAIN VALUE"

        val cdrVersion = CdrVersion()
        cdrVersionDao!!.save(cdrVersion)

        val dataDictionaryEntry = DataDictionaryEntry()
        dataDictionaryEntry.cdrVersion = cdrVersion
        dataDictionaryEntry.definedTime = Timestamp(CLOCK.millis())
        dataDictionaryEntry.relevantOmopTable = ConceptSetDao.DOMAIN_TO_TABLE_NAME[domain]
        dataDictionaryEntry.fieldName = domainValue
        dataDictionaryEntry.omopCdmStandardOrCustomField = "A"
        dataDictionaryEntry.description = "B"
        dataDictionaryEntry.fieldType = "C"
        dataDictionaryEntry.dataProvenance = "D"
        dataDictionaryEntry.sourcePpiModule = "E"
        dataDictionaryEntry.transformedByRegisteredTierPrivacyMethods = true

        dataDictionaryEntryDao!!.save(dataDictionaryEntry)

        val response = dataSetController!!
                .getDataDictionaryEntry(cdrVersion.cdrVersionId, domain.toString(), domainValue)
                .body

        assertThat(response.getCdrVersionId().longValue())
                .isEqualTo(dataDictionaryEntry.cdrVersion!!.cdrVersionId)
        assertThat(Timestamp(response.getDefinedTime()))
                .isEqualTo(dataDictionaryEntry.definedTime)
        assertThat(response.getRelevantOmopTable())
                .isEqualTo(dataDictionaryEntry.relevantOmopTable)
        assertThat(response.getFieldName()).isEqualTo(dataDictionaryEntry.fieldName)
        assertThat(response.getOmopCdmStandardOrCustomField())
                .isEqualTo(dataDictionaryEntry.omopCdmStandardOrCustomField)
        assertThat(response.getDescription()).isEqualTo(dataDictionaryEntry.description)
        assertThat(response.getFieldType()).isEqualTo(dataDictionaryEntry.fieldType)
        assertThat(response.getDataProvenance()).isEqualTo(dataDictionaryEntry.dataProvenance)
        assertThat(response.getSourcePpiModule()).isEqualTo(dataDictionaryEntry.sourcePpiModule)
        assertThat(response.getTransformedByRegisteredTierPrivacyMethods())
                .isEqualTo(dataDictionaryEntry.transformedByRegisteredTierPrivacyMethods)
    }

    @Test
    fun testGetDataDictionaryEntry_invalidCdr() {
        expectedEx.expect(BadRequestException::class.java)
        expectedEx.expectMessage("Invalid CDR Version")

        dataSetController!!.getDataDictionaryEntry(-1L, Domain.DRUG.toString(), "TEST FIELD")
    }

    @Test
    fun testGetDataDictionaryEntry_invalidDomain() {
        expectedEx.expect(BadRequestException::class.java)
        expectedEx.expectMessage("Invalid Domain")

        dataSetController!!.getDataDictionaryEntry(
                cdrVersionDao!!.findAll().iterator().next().cdrVersionId, "random", "TEST FIELD")
    }

    @Test
    fun testGetDataDictionaryEntry_notFound() {
        expectedEx.expect(NotFoundException::class.java)

        dataSetController!!.getDataDictionaryEntry(1L, Domain.DRUG.toString(), "random")
    }

    companion object {

        private val NOW = Instant.now()
        private val CLOCK = FakeClock(NOW, ZoneId.systemDefault())
    }
}
